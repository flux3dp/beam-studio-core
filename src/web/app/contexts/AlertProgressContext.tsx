import React, { createContext } from 'react';

import AlertConstants from 'app/constants/alert-constants';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import i18n from 'helpers/i18n';
import ProgressConstants from 'app/constants/progress-constants';
import { IAlert } from 'interfaces/IAlert';
import { IProgressDialog } from 'interfaces/IProgress';

const LANG = i18n.lang.alert;
let progressID = 0;

interface IAlertProgressContext {
  alertProgressStack: (IAlert | IProgressDialog)[];
  popFromStack: () => void;
  popById: (id: string) => void;
}

export const AlertProgressContext = createContext<IAlertProgressContext>({
  alertProgressStack: [],
  popFromStack: () => {},
  popById: () => {},
});

export const eventEmitter = eventEmitterFactory.createEventEmitter();

interface State {
  alertProgressStack: (IAlert | IProgressDialog)[],
}

export class AlertProgressContextProvider extends React.Component<unknown, State> {
  constructor(props: unknown) {
    super(props);
    this.state = {
      alertProgressStack: [],
    };

    eventEmitter.on('OPEN_PROGRESS', this.openProgress.bind(this));
    eventEmitter.on('POP_LAST_PROGRESS', this.popLastProgress.bind(this));
    eventEmitter.on('UPDATE_PROGRESS', this.updateProgress.bind(this));
    eventEmitter.on('POP_BY_ID', this.popById.bind(this));
    eventEmitter.on('POP_UP', this.popUp.bind(this));
    eventEmitter.on('CHECK_ID_EXIST', this.checkIdExist.bind(this));
  }

  componentWillUnmount() {
    eventEmitter.removeAllListeners();
  }

  popFromStack = (): void => {
    const { alertProgressStack } = this.state;
    alertProgressStack.pop();
    this.forceUpdate();
  };

  popById = (id: string): void => {
    const { alertProgressStack } = this.state;
    this.setState({
      alertProgressStack: alertProgressStack.filter((item) => item.id !== id),
    });
  };

  checkIdExist = (id: string, response: {
    idExist: boolean,
  }): void => {
    const { alertProgressStack } = this.state;
    const res = alertProgressStack.filter((item) => {
      const { id: itemId } = item;
      return itemId === id && !('isProgress' in item);
    });
    response.idExist = res.length > 0;
  };

  pushToStack = (item: (IAlert | IProgressDialog)): void => {
    if (item.id) {
      // eslint-disable-next-line no-console
      console.log('alert/progress poped', item.id);
    }
    const { alertProgressStack } = this.state;
    alertProgressStack.push(item);
    this.forceUpdate();
  };

  openProgress = (args: IProgressDialog): void => {
    const { type } = args;
    let { id, message, caption } = args;
    if (!id) {
      id = `progress_${progressID}`;
      progressID = (progressID + 1) % 100000;
    }
    message = message || '';
    caption = caption || '';

    this.pushToStack({
      ...args,
      id,
      type,
      caption,
      message,
      isProgress: true,
    });
  };

  popLastProgress = (): void => {
    const { alertProgressStack } = this.state;
    let i;
    for (i = alertProgressStack.length - 1; i >= 0; i -= 1) {
      if ('isProgress' in alertProgressStack[i]) {
        break;
      }
    }
    if (i >= 0) {
      alertProgressStack.splice(i, 1);
      this.setState({ alertProgressStack });
    }
  };

  updateProgress = (id: string, args: IProgressDialog): void => {
    const { alertProgressStack } = this.state;
    const targetObjects = alertProgressStack.filter((item) => {
      const { id: itemId } = item;
      return 'isProgress' in item && itemId === id;
    });
    if (targetObjects.length === 0) {
      return;
    }
    const targetObject = targetObjects[targetObjects.length - 1];
    if (targetObject.type === ProgressConstants.NONSTOP && !args.caption && args.message) {
      // eslint-disable-next-line no-param-reassign
      args.caption = args.message;
    }
    Object.assign(targetObject, args);
    this.setState({ alertProgressStack });
  };

  popUp = (args: IAlert): void => {
    const { type } = args;
    let { message, caption } = args;
    message = message || '';
    switch (type) {
      case AlertConstants.SHOW_POPUP_INFO:
        caption = caption || LANG.info;
        break;
      case AlertConstants.SHOW_POPUP_WARNING:
        caption = caption || LANG.warning;
        break;
      case AlertConstants.SHOW_POPUP_ERROR:
        caption = caption || LANG.error;
        break;
      default:
        break;
    }
    const { buttons, checkbox } = this.buttonsGenerator(args);
    const checkboxText = checkbox ? checkbox.text : null;
    const checkboxCallbacks = checkbox ? checkbox.callbacks : null;

    this.pushToStack({
      ...args,
      caption,
      message,
      buttons,
      checkboxText,
      checkboxCallbacks,
    });
  };

  buttonsGenerator = (args: IAlert): { buttons, checkbox } => {
    const { checkbox } = args;
    let { buttons } = args;
    if (buttons) {
      return { buttons, checkbox };
    }
    const {
      id,
      buttonType,
    } = args;
    let {
      buttonLabels,
      callbacks,
      primaryButtonIndex,
      onYes,
      onNo,
      onConfirm,
      onRetry,
      onCancel,
    } = args;
    switch (buttonType) {
      case AlertConstants.YES_NO:
        onYes = onYes || (() => { });
        onNo = onNo || (() => { });
        buttonLabels = [LANG.yes, LANG.no];
        callbacks = [onYes, onNo];
        primaryButtonIndex = primaryButtonIndex || 0;
        if (checkbox) {
          const onCheckedYes = checkbox.onYes ? checkbox.onYes : () => { };
          const onCheckedNo = checkbox.onNo ? checkbox.onNo : () => { };
          checkbox.callbacks = [onCheckedYes, onCheckedNo];
        }
        break;
      case AlertConstants.YES_NO_CUSTOM:
        onYes = onYes || (() => { });
        onNo = onNo || (() => { });
        primaryButtonIndex = primaryButtonIndex || 0;
        if (!buttonLabels) {
          buttonLabels = [LANG.yes, LANG.no];
          callbacks = [onYes, onNo];
          if (checkbox) {
            const onCheckedYes = checkbox.onYes ? checkbox.onYes : () => { };
            const onCheckedNo = checkbox.onNo ? checkbox.onNo : () => { };
            checkbox.callbacks = [onCheckedYes, onCheckedNo];
          }
        } else if (typeof buttonLabels === 'string') {
          buttonLabels = [LANG.yes, LANG.no, buttonLabels];
          callbacks = callbacks || (() => { });
          callbacks = [onYes, onNo, (callbacks as () => void)];
        } else {
          buttonLabels = [LANG.yes, LANG.no, ...buttonLabels];
          callbacks = [onYes, onNo, ...(callbacks as (() => void)[])];
        }
        break;
      case AlertConstants.CONFIRM_CANCEL:
        onConfirm = onConfirm || (() => { });
        onCancel = onCancel || (() => { });
        buttonLabels = [LANG.confirm, LANG.cancel];
        primaryButtonIndex = primaryButtonIndex || 0;
        callbacks = [onConfirm, onCancel];
        if (checkbox) {
          const onCheckedConfirm = checkbox.onConfirm ? checkbox.onConfirm : () => { };
          const onCheckedCancel = checkbox.onCancel ? checkbox.onCancel : () => { };
          checkbox.callbacks = [onCheckedConfirm, onCheckedCancel];
        }
        break;
      case AlertConstants.RETRY_CANCEL:
        onRetry = onRetry || (() => { });
        onCancel = onCancel || (() => { });
        buttonLabels = [LANG.retry, LANG.cancel];
        primaryButtonIndex = primaryButtonIndex || 0;
        callbacks = [onRetry, onCancel];
        if (checkbox) {
          const onCheckedRetry = checkbox.onRetry ? checkbox.onRetry : () => { };
          const onCheckedCancel = checkbox.onCancel ? checkbox.onCancel : () => { };
          checkbox.callbacks = [onCheckedRetry, onCheckedCancel];
        }
        break;
      case AlertConstants.CUSTOM_CANCEL:
        onCancel = onCancel || (() => { });
        primaryButtonIndex = primaryButtonIndex || 0;
        if (!buttonLabels) {
          buttonLabels = [LANG.cancel];
          callbacks = [onCancel];
          if (checkbox) {
            const onCheckedCancel = checkbox.onCancel ? checkbox.onCancel : () => { };
            checkbox.callbacks = [onCheckedCancel];
          }
        } else if (typeof buttonLabels === 'string') {
          buttonLabels = [buttonLabels, LANG.cancel];
          callbacks = callbacks || (() => { });
          callbacks = [callbacks as () => void, onCancel];
        } else {
          buttonLabels = [...buttonLabels, LANG.cancel];
          callbacks = [...(callbacks as (() => void)[]), onCancel];
        }
        break;
      default:
        if (!buttonLabels) {
          buttonLabels = [LANG.ok];
          callbacks = callbacks || (() => { });
          if (checkbox) {
            if (!checkbox.callbacks) {
              checkbox.callbacks = [() => { }];
            } else if (typeof checkbox.callbacks === 'function') {
              checkbox.callbacks = [checkbox.callbacks];
            }
          }
        } else if (typeof buttonLabels === 'string') {
          buttonLabels = [buttonLabels];
          callbacks = callbacks || (() => { });
          if (checkbox) {
            if (!checkbox.callbacks) {
              checkbox.callbacks = [() => { }];
            } else if (typeof checkbox.callbacks === 'function') {
              checkbox.callbacks = [checkbox.callbacks];
            }
          }
        }
        break;
    }
    buttons = buttonLabels.map((label, i) => {
      const b = {
        label,
        className: (buttonLabels.length === 1 || i === primaryButtonIndex || primaryButtonIndex === undefined) ? 'btn-default primary' : 'btn-default',
        onClick: () => { },
      };
      if (callbacks && typeof callbacks === 'function') {
        b.onClick = () => {
          (callbacks as (id: string) => void)(id);
        };
      } else if (callbacks && callbacks.length > i) {
        b.onClick = () => {
          callbacks[i](id);
        };
      } else if (!callbacks) {
        b.onClick = () => { };
      }
      return b;
    });

    return { buttons, checkbox };
  };

  render(): JSX.Element {
    const { children } = this.props;
    const { alertProgressStack } = this.state;
    return (
      <AlertProgressContext.Provider value={
          {
            alertProgressStack,
            popFromStack: this.popFromStack,
            popById: this.popById,
          }
        }
      >
        {children}
      </AlertProgressContext.Provider>
    );
  }
}
