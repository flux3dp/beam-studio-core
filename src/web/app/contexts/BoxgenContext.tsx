import React, { createContext, Dispatch, SetStateAction, useMemo, useState } from 'react';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import constant, { WorkareaMap } from 'app/actions/beambox/constant';
import LayerModule from 'app/constants/layer-module/layer-modules';
import layerModuleHelper from 'helpers/layer-module/layer-module-helper';
import moduleBoundary from 'app/constants/layer-module/module-boundary';
import storage from 'implementations/storage';
import { DEFAULT_CONTROLLER_INCH, DEFAULT_CONTROLLER_MM } from 'app/constants/boxgen-constants';
import { IController } from 'interfaces/IBoxgen';

interface BoxgenContextType {
  onClose: () => void;
  boxData: IController;
  setBoxData: Dispatch<SetStateAction<IController>>;
  zoomKey: number;
  setZoomKey: Dispatch<SetStateAction<number>>;
  resetKey: number;
  setResetKey: Dispatch<SetStateAction<number>>;
  workarea: { value: string; label: string; canvasWidth: number; canvasHeight: number };
  lengthUnit: { unit: 'mm' | 'inch'; unitRatio: number; decimal: number };
}

export const BoxgenContext = createContext<BoxgenContextType>({
  onClose: () => {},
  boxData: DEFAULT_CONTROLLER_MM,
  setBoxData: () => {},
  zoomKey: 0,
  setZoomKey: () => {},
  resetKey: 0,
  setResetKey: () => {},
  workarea: { value: 'fbm1', label: 'beamo', canvasWidth: 300, canvasHeight: 210 },
  lengthUnit: { unit: 'mm', unitRatio: 1, decimal: 0 },
});

interface BoxgenProviderProps {
  onClose: () => void;
  children: React.ReactNode;
}

export function BoxgenProvider({ onClose, children }: BoxgenProviderProps): JSX.Element {
  const workareaValue = BeamboxPreference.read('workarea') || 'fbm1';
  const workarea = useMemo(() => {
    const currentWorkarea = WorkareaMap.get(workareaValue);
    if (currentWorkarea.label === 'Ador') {
      const laserModule = layerModuleHelper.getDefaultLaserModule();
      const boundary = moduleBoundary[laserModule] as {
        top: number;
        left: number;
        right: number;
        bottom: number;
      };
      return {
        value: workareaValue,
        label: `${currentWorkarea.label} ${
          laserModule === LayerModule.LASER_10W_DIODE ? '10W' : '20W'
        }`,
        canvasWidth: currentWorkarea.width / constant.dpmm - boundary.left - boundary.right,
        canvasHeight: currentWorkarea.height / constant.dpmm - boundary.top - boundary.bottom,
      };
    }
    return {
      value: workareaValue,
      label: currentWorkarea.label,
      canvasWidth: currentWorkarea.width / constant.dpmm,
      canvasHeight: currentWorkarea.height / constant.dpmm,
    };
  }, [workareaValue]);

  const unit = storage.get('default-units') || 'mm';
  const isMM = unit === 'mm';
  const lengthUnit = isMM
    ? { unit: 'mm' as const, unitRatio: 1, decimal: 0 }
    : { unit: 'inch' as const, unitRatio: 25.4, decimal: 3 };

  const [boxData, setBoxData] = useState(isMM ? DEFAULT_CONTROLLER_MM : DEFAULT_CONTROLLER_INCH);
  const [zoomKey, setZoomKey] = useState(0);
  const [resetKey, setResetKey] = useState(0);

  return (
    <BoxgenContext.Provider
      value={{
        onClose,
        boxData,
        setBoxData,
        zoomKey,
        setZoomKey,
        resetKey,
        setResetKey,
        workarea,
        lengthUnit,
      }}
    >
      {children}
    </BoxgenContext.Provider>
  );
}
