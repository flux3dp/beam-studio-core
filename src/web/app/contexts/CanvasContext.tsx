import React, { createContext, useCallback, useEffect, useState } from 'react';

import beamboxPreference from 'app/actions/beambox/beambox-preference';
import curveEngravingModeController from 'app/actions/canvas/curveEngravingModeController';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import FnWrapper from 'app/actions/beambox/svgeditor-function-wrapper';
import PreviewModeController from 'app/actions/beambox/preview-mode-controller';
import TutorialConstants from 'app/constants/tutorial-constants';
import useForceUpdate from 'helpers/use-force-update';
import workareaManager from 'app/svgedit/workarea';
import * as TutorialController from 'app/views/tutorials/tutorialController';
import { getSupportInfo } from 'app/constants/add-on';
import { getSVGAsync } from 'helpers/svg-editor-helper';
import { IDeviceInfo } from 'interfaces/IDevice';
import { IUser } from 'interfaces/IUser';

const canvasEventEmitter = eventEmitterFactory.createEventEmitter('canvas');
const topBarEventEmitter = eventEmitterFactory.createEventEmitter('top-bar');
const fluxIDEventEmitter = eventEmitterFactory.createEventEmitter('flux-id');
const workareaEventEmitter = eventEmitterFactory.createEventEmitter('workarea');

let svgCanvas;
getSVGAsync((globalSVG) => {
  svgCanvas = globalSVG.Canvas;
});

const workareaEvents = eventEmitterFactory.createEventEmitter('workarea');

export enum CanvasMode {
  Draw = 1,
  Preview = 2,
  PathPreview = 3,
  CurveEngraving = 4,
}

interface CanvasContextType {
  changeToPreviewMode: () => void;
  currentUser: IUser;
  endPreviewMode: () => void;
  hasUnsavedChange: boolean;
  mode: CanvasMode;
  setMode: (mode: CanvasMode) => void;
  setShouldStartPreviewController: (shouldStartPreviewController: boolean) => void;
  setStartPreviewCallback: (callback: () => void | null) => void;
  shouldStartPreviewController: boolean;
  setupPreviewMode: () => void;
  setSetupPreviewMode: (callback: () => void | null) => void;
  startPreviewCallback: () => void | null;
  togglePathPreview: () => void;
  updateCanvasContext: () => void;
  selectedDevice: IDeviceInfo | null;
  setSelectedDevice: (IDeviceInfo) => void;
  isColorPreviewing: boolean;
  setIsColorPreviewing: (isColorPreviewing: boolean) => void;
  isPathEditing: boolean;
  setIsPathEditing: (isPathEditing: boolean) => void;
  hasPassthroughExtension: boolean;
}

const CanvasContext = createContext<CanvasContextType>({
  changeToPreviewMode: () => {},
  currentUser: null,
  endPreviewMode: () => {},
  hasUnsavedChange: false,
  mode: CanvasMode.Draw,
  setMode: () => {},
  setShouldStartPreviewController: () => {},
  setStartPreviewCallback: () => {},
  shouldStartPreviewController: false,
  setupPreviewMode: () => {},
  setSetupPreviewMode: () => {},
  startPreviewCallback: () => {},
  togglePathPreview: () => {},
  updateCanvasContext: () => {},
  selectedDevice: null,
  setSelectedDevice: () => {},
  isColorPreviewing: false,
  setIsColorPreviewing: () => {},
  isPathEditing: false,
  setIsPathEditing: () => {},
  hasPassthroughExtension: false,
});

const CanvasProvider = (props: React.PropsWithChildren<Record<string, unknown>>): JSX.Element => {
  const forceUpdate = useForceUpdate();
  const [mode, setMode] = useState<CanvasMode>(CanvasMode.Draw);
  const [isColorPreviewing, setIsColorPreviewing] = useState<boolean>(false);
  const [currentUser, setCurrentUser] = useState<IUser>(null);
  const [hasUnsavedChange, setHasUnsavedChange] = useState<boolean>(false);
  const [startPreviewCallback, setStartPreviewCallback] = useState<() => void | null>(null);
  const [shouldStartPreviewController, setShouldStartPreviewController] = useState<boolean>(false);
  const [setupPreviewMode, setSetupPreviewMode] = useState<() => void | null>(null);
  const [selectedDevice, setSelectedDevice] = useState<IDeviceInfo | null>(null);
  const [isPathEditing, setIsPathEditing] = useState<boolean>(false);
  const [hasPassthroughExtension, setHasPassthroughExtension] = useState<boolean>(false);

  const endPreviewMode = (): void => {
    try {
      if (PreviewModeController.isPreviewMode()) {
        PreviewModeController.end();
      }
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log(error);
    } finally {
      if (TutorialController.getNextStepRequirement() === TutorialConstants.TO_EDIT_MODE) {
        TutorialController.handleNextStep();
      }
      // eslint-disable-next-line react-hooks/rules-of-hooks
      FnWrapper.useSelectTool();
      $('#workarea').off('contextmenu');
      workareaEventEmitter.emit('update-context-menu', { menuDisabled: false });
      setMode(CanvasMode.Draw);
    }
  };

  const setUser = useCallback((user) => setCurrentUser({ ...user }), []);
  useEffect(() => {
    fluxIDEventEmitter.on('update-user', setUser);
    const handler = (e: CustomEvent) => {
      setUser(e.detail.user);
    };
    window.addEventListener('update-user', handler);
    return () => {
      fluxIDEventEmitter.removeListener('update-user', setUser);
      window.removeEventListener('update-user', handler);
    };
  }, [setUser]);
  useEffect(() => {
    topBarEventEmitter.on('SET_HAS_UNSAVED_CHANGE', setHasUnsavedChange);
    topBarEventEmitter.on('SET_SHOULD_START_PREVIEW_CONTROLLER', setShouldStartPreviewController);
    topBarEventEmitter.on('SET_SELECTED_DEVICE', setSelectedDevice);
    const setStartPreviewCallbackHandler = (callback) => {
      // wrap callback with a function to prevent calling it immediately
      setStartPreviewCallback(() => callback);
    };
    topBarEventEmitter.on('SET_START_PREVIEW_CALLBACK', setStartPreviewCallbackHandler);
    return () => {
      topBarEventEmitter.removeListener('SET_HAS_UNSAVED_CHANGE', setHasUnsavedChange);
      topBarEventEmitter.removeListener(
        'SET_SHOULD_START_PREVIEW_CONTROLLER',
        setShouldStartPreviewController
      );
      topBarEventEmitter.removeListener('SET_SELECTED_DEVICE', setSelectedDevice);
      topBarEventEmitter.removeListener(
        'SET_START_PREVIEW_CALLBACK',
        setStartPreviewCallbackHandler
      );
    };
  }, []);
  useEffect(() => {
    const handler = (response: { isPreviewMode: boolean }): void => {
      response.isPreviewMode = mode === CanvasMode.Preview;
    };
    topBarEventEmitter.on('GET_TOP_BAR_PREVIEW_MODE', handler);
    return () => {
      topBarEventEmitter.removeListener('GET_TOP_BAR_PREVIEW_MODE', handler);
    };
  }, [mode]);
  useEffect(() => {
    const handler = (response: { selectedDevice: IDeviceInfo | null }): void => {
      response.selectedDevice = selectedDevice;
    };
    topBarEventEmitter.on('GET_SELECTED_DEVICE', handler);
    return () => {
      topBarEventEmitter.removeListener('GET_SELECTED_DEVICE', handler);
    };
  }, [selectedDevice]);

  const updateCanvasContext = useCallback(() => {
    forceUpdate();
  }, [forceUpdate]);

  useEffect(() => {
    canvasEventEmitter.on('UPDATE_CONTEXT', updateCanvasContext); // This force rerender the context
    return () => {
      canvasEventEmitter.removeListener('UPDATE_CONTEXT', updateCanvasContext);
    };
  }, [updateCanvasContext]);

  useEffect(() => {
    canvasEventEmitter.on('SET_COLOR_PREVIEWING', setIsColorPreviewing);
    canvasEventEmitter.on('SET_PATH_EDITING', setIsPathEditing);
    canvasEventEmitter.on('SET_MODE', setMode);
    const canvasChangeHandler = () =>
      setHasPassthroughExtension(
        beamboxPreference.read('pass-through') && getSupportInfo(workareaManager.model).passThrough
      );
    canvasChangeHandler();
    canvasEventEmitter.on('canvas-change', canvasChangeHandler);
    return () => {
      canvasEventEmitter.removeListener('SET_COLOR_PREVIEWING', setIsColorPreviewing);
      canvasEventEmitter.removeListener('SET_PATH_EDITING', setIsPathEditing);
      canvasEventEmitter.removeListener('SET_MODE', setMode);
      canvasEventEmitter.removeListener('canvas-change', canvasChangeHandler);
    };
  }, []);

  useEffect(() => {
    if (mode !== CanvasMode.CurveEngraving && curveEngravingModeController.started) {
      curveEngravingModeController.end();
    }
    const allLayers = document.querySelectorAll('g.layer');
    for (let i = 0; i < allLayers.length; i += 1) {
      const g = allLayers[i] as SVGGElement;
      if (mode === CanvasMode.Preview) {
        g.style.pointerEvents = 'none';
      } else {
        g.style.pointerEvents = '';
      }
    }
  }, [mode]);

  const changeToPreviewMode = () => {
    svgCanvas.setMode('select');
    workareaEvents.emit('update-context-menu', { menuDisabled: true });
    const workarea = document.getElementById('workarea');
    $('#workarea').contextmenu(() => {
      endPreviewMode();
      return false;
    });
    setMode(CanvasMode.Preview);
    if (workarea) {
      workarea.style.cursor = 'url(img/camera-cursor.svg), cell';
    }
    if (TutorialController.getNextStepRequirement() === TutorialConstants.TO_PREVIEW_MODE) {
      TutorialController.handleNextStep();
    }
  };

  const togglePathPreview = () => {
    setMode(mode === CanvasMode.PathPreview ? CanvasMode.Draw : CanvasMode.PathPreview);
  };

  const { children } = props;
  return (
    <CanvasContext.Provider
      value={{
        changeToPreviewMode,
        currentUser,
        endPreviewMode,
        hasUnsavedChange,
        mode,
        setMode,
        setShouldStartPreviewController,
        setSetupPreviewMode,
        setStartPreviewCallback,
        shouldStartPreviewController,
        setupPreviewMode,
        startPreviewCallback,
        togglePathPreview,
        updateCanvasContext,
        selectedDevice,
        setSelectedDevice,
        isColorPreviewing,
        setIsColorPreviewing,
        isPathEditing,
        setIsPathEditing,
        hasPassthroughExtension,
      }}
    >
      {children}
    </CanvasContext.Provider>
  );
};

export { CanvasContextType, CanvasContext, CanvasProvider };
