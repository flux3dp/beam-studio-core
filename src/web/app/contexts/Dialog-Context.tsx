import * as React from 'react';

const { createContext } = React;
export const DialogContext = createContext(null);

export class DialogContextProvider extends React.Component<any> {
  private dialogComponents: {
    id: string,
    component: JSX.Element
  }[];

  constructor(props) {
    super(props);
    this.dialogComponents = [];
  }

  addDialogComponent = (id: string, dialogComponent: JSX.Element): void => {
    this.dialogComponents.push({ id, component: dialogComponent });
    this.forceUpdate();
  };

  isIdExist = (id: string): boolean => this.dialogComponents.some((dialog) => dialog.id === id);

  popDialogById = (id: string): void => {
    this.dialogComponents = this.dialogComponents.filter((dialog) => dialog.id !== id);
    this.forceUpdate();
  };

  clearAllDialogComponents = (): void => {
    this.dialogComponents = [];
    this.forceUpdate();
  };

  render() {
    const { children } = this.props;
    const {
      dialogComponents,
      addDialogComponent,
      clearAllDialogComponents,
      isIdExist,
      popDialogById,
    } = this;

    return (
      <DialogContext.Provider
        value={{
          dialogComponents,
          addDialogComponent,
          clearAllDialogComponents,
          isIdExist,
          popDialogById,
        }}
      >
        {children}
      </DialogContext.Provider>
    );
  }
}
