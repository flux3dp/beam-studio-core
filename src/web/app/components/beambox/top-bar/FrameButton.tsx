import classNames from 'classnames';
import React, { useContext, useEffect } from 'react';

import shortcuts from 'helpers/shortcuts';
import TopBarController from 'app/views/beambox/TopBar/contexts/TopBarController';
import TopBarIcons from 'app/icons/top-bar/TopBarIcons';
import useI18n from 'helpers/useI18n';
import { CanvasContext, CanvasMode } from 'app/contexts/CanvasContext';
import { promarkModels } from 'app/actions/beambox/constant';
import { showFramingModal } from 'app/components/dialogs/FramingModal';

import styles from './FrameButton.module.scss';

const FrameButton = (): JSX.Element => {
  const lang = useI18n();
  const { mode } = useContext(CanvasContext);
  useEffect(() => {
    const shortcutHandler = async () => {
      if (mode !== CanvasMode.Draw) return;
      const device = TopBarController.getSelectedDevice();
      showFramingModal(promarkModels.has(device?.model));
    };
    const unregister = shortcuts.on(['F1'], shortcutHandler);
    return () => unregister?.();
  }, [mode]);

  return (
    <div
      className={classNames(styles.button, { [styles.disabled]: mode !== CanvasMode.Draw })}
      onClick={() => showFramingModal()}
      title={lang.topbar.frame_task}
    >
      <TopBarIcons.Frame />
    </div>
  );
};

export default FrameButton;
