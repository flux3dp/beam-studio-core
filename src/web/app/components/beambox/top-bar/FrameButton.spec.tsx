import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { CanvasContext } from 'app/contexts/CanvasContext';

import FrameButton from './FrameButton';

const mockGetVisibleElementsAndBBoxes = jest
  .fn()
  .mockReturnValue([{ bbox: { x: 0, y: 1, width: 2, height: 3 } }]);
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) =>
    callback({
      Canvas: {
        getVisibleElementsAndBBoxes: (...args) => mockGetVisibleElementsAndBBoxes(...args),
      },
    }),
}));

const beamboxPreferenceRead = jest.fn();
jest.mock('app/actions/beambox/beambox-preference', () => ({
  read: (...args) => beamboxPreferenceRead(...args),
}));

const mockOpenMessage = jest.fn();
jest.mock('app/actions/message-caller', () => ({
  __esModule: true,
  default: { openMessage: (...args) => mockOpenMessage(...args) },
  MessageLevel: {},
}));

const mockOpenNonstopProgress = jest.fn();
const mockUpdate = jest.fn();
const mockPopById = jest.fn();

jest.mock('app/actions/progress-caller', () => ({
  openNonstopProgress: (...args) => mockOpenNonstopProgress(...args),
  update: (...args) => mockUpdate(...args),
  popById: (...args) => mockPopById(...args),
}));

jest.mock('helpers/useI18n', () => () => ({
  message: {
    connectingMachine: 'Connecting %s...',
    enteringRawMode: 'Entering raw mode...',
    exitingRotaryMode: 'Exiting rotary mode...',
    homing: 'Homing...',
    turningOffFan: 'Turning off fan...',
    turningOffAirPump: 'Turning off air pump...',
  },
  topbar: {
    frame_task: 'frame_task',
    alerts: {
      add_content_first: 'Please add objects first',
    }
  },
  device: {
    processing: 'Processing',
  },
}));

jest.mock('app/contexts/CanvasContext', () => ({
  CanvasContext: React.createContext({ isPreviewing: false }),
}));

const versionChecker = jest.fn().mockReturnValue({ meetRequirement: () => true });
jest.mock(
  'helpers/version-checker',
  () =>
    (...args) =>
      versionChecker(...args)
);

const mockCheckDeviceStatus = jest.fn().mockResolvedValue({});
jest.mock(
  'helpers/check-device-status',
  () =>
    (...args) =>
      mockCheckDeviceStatus(...args)
);

const mockEnterRawMode = jest.fn();
const mockRawSetRotary = jest.fn();
const mockRawHome = jest.fn();
const mockRawStartLineCheckMode = jest.fn();
const mockRawSetFan = jest.fn();
const mockRawSetAirPump = jest.fn();
const mockRawSetWaterPump = jest.fn();
const mockRawMove = jest.fn();
const mockRawEndLineCheckMode = jest.fn();
const mockRawLooseMotor = jest.fn();
const mockEndRawMode = jest.fn();
const mockKick = jest.fn();

jest.mock('helpers/device-master', () => ({
  enterRawMode: (...args) => mockEnterRawMode(...args),
  rawSetRotary: (...args) => mockRawSetRotary(...args),
  rawHome: (...args) => mockRawHome(...args),
  rawStartLineCheckMode: (...args) => mockRawStartLineCheckMode(...args),
  rawSetFan: (...args) => mockRawSetFan(...args),
  rawSetAirPump: (...args) => mockRawSetAirPump(...args),
  rawSetWaterPump: (...args) => mockRawSetWaterPump(...args),
  rawMove: (...args) => mockRawMove(...args),
  rawEndLineCheckMode: (...args) => mockRawEndLineCheckMode(...args),
  rawLooseMotor: (...args) => mockRawLooseMotor(...args),
  endRawMode: (...args) => mockEndRawMode(...args),
  kick: (...args) => mockKick(...args),
}));

const mockGetDevice = jest.fn().mockResolvedValue({ device: {} });
jest.mock(
  'helpers/device/get-device',
  () =>
    (...args) =>
      mockGetDevice(...args)
);

describe('test FrameButton', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('should render correctly', async () => {
    const { container } = render(<FrameButton />);
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('div[class*="button"]'));
    await waitFor(() => expect(mockPopById).toBeCalledTimes(1));
    expect(mockGetDevice).toBeCalledTimes(1);
    expect(mockCheckDeviceStatus).toBeCalledTimes(1);
    expect(mockOpenNonstopProgress).toBeCalledTimes(1);
    expect(mockUpdate).toBeCalledTimes(6);
    expect(mockEnterRawMode).toBeCalledTimes(1);
    expect(mockRawSetRotary).toBeCalledTimes(1);
    expect(mockRawHome).toBeCalledTimes(1);
    expect(mockRawStartLineCheckMode).toBeCalledTimes(1);
    expect(mockRawSetFan).toBeCalledTimes(1);
    expect(mockRawSetAirPump).toBeCalledTimes(1);
    expect(mockRawSetWaterPump).toBeCalledTimes(1);
    expect(mockRawMove).toBeCalledTimes(5);
    expect(mockRawEndLineCheckMode).toBeCalledTimes(1);
    expect(mockRawLooseMotor).toBeCalledTimes(1);
    expect(mockEndRawMode).toBeCalledTimes(1);
    expect(mockKick).toBeCalledTimes(1);
  });

  test('should render correctly with previewing mode', () => {
    const { container } = render(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      <CanvasContext.Provider value={{ isPreviewing: true } as any}>
        <FrameButton />
      </CanvasContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('no element', () => {
    mockGetVisibleElementsAndBBoxes.mockReturnValue([]);
    const { container } = render(<FrameButton />);
    fireEvent.click(container.querySelector('div[class*="button"]'));
    expect(mockOpenMessage).toBeCalledTimes(1);
  });
});
