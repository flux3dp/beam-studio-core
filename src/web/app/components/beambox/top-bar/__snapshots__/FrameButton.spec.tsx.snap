// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test FrameButton should render correctly 1`] = `
<div>
  <div
    class="button"
  >
    <svgr-url />
  </div>
</div>
`;

exports[`test FrameButton should render correctly with previewing mode 1`] = `
<div>
  <div
    class="button disabled"
  >
    <svgr-url />
  </div>
</div>
`;
